kythe/cxx/tools/kindex_tool_main.cc

```diff
@@ -83,6 +86,59 @@ void CanonicalizeHash(std::map<google::protobuf::string, size_t>* hashes,
       google::protobuf::string("hash" + std::to_string(inserted.first->second));
 }
 
+void DumpCompilationUnit(const std::string& path,
+                         kythe::proto::CompilationUnit* unit) {
+  using namespace google::protobuf::io;
+  std::map<google::protobuf::string, size_t> hash_table;
+  std::string out_path = path + "_UNIT";
+  int out_fd =
+      open(out_path.c_str(), O_WRONLY | O_CREAT | O_TRUNC, S_IREAD | S_IWRITE);
+  CHECK_GE(out_fd, 0) << "Couldn't open " << out_path << " for writing.";
+  if (FLAGS_suppress_details) {
+    unit->clear_details();
+  }
+  if (FLAGS_canonicalize_hashes) {
+    CanonicalizeHash(&hash_table, unit->mutable_entry_context());
+    for (auto& input : *unit->mutable_required_input()) {
+      for (auto& row : MutableContextRows(&input)) {
+        CanonicalizeHash(&hash_table, row.mutable_source_context());
+        for (auto& column : *row.mutable_column()) {
+          CanonicalizeHash(&hash_table, column.mutable_linked_context());
+        }
+      }
+      // TODO(shahms): Remove this when the indexers are updated.
+      if (input.has_context()) {
+        for (auto& row : *input.mutable_context()->mutable_row()) {
+          CanonicalizeHash(&hash_table, row.mutable_source_context());
+          for (auto& column : *row.mutable_column()) {
+            CanonicalizeHash(&hash_table, column.mutable_linked_context());
+          }
+        }
+      }
+    }
+  }
+  FileOutputStream file_output_stream(out_fd);
+  google::protobuf::TextFormat::Printer printer;
+  printer.SetExpandAny(true);
+  CHECK(printer.Print(*unit, &file_output_stream));
+  CHECK(file_output_stream.Close());
+}
+
+void DumpFileData(const std::string& path,
+                  const kythe::proto::FileData& content) {
+  using namespace google::protobuf::io;
+  CHECK(content.has_info() && !content.info().digest().empty());
+  std::string out_path = path + "_" + content.info().digest();
+  int out_fd =
+      open(out_path.c_str(), O_WRONLY | O_CREAT | O_TRUNC, S_IREAD | S_IWRITE);
+  CHECK_GE(out_fd, 0) << "Couldn't open " << out_path << " for writing.";
+  FileOutputStream file_output_stream(out_fd);
```

Comment:
I think you need to tell the FileOutputStream to close the FD when it's done? (Same above)

Reply:
It's closed on line 139. Above, on 124.

Reply:
Ah, fair enough. (I was confusing the Close() semantics with SetCloseOnDelete())